### PR TITLE
[sonic-buildimage] New feature managementVRF(L3mdev)

### PR DESCRIPTION
**Pull request by vharish02**
This commit adds support for New feature management VRF using L3mdev.
Added commands to enable/disable management VRF. Config vrf add mgmt
will enable management VRF, enslave the eth0 device to the master
device mgmt, stop ntp service in default, restart interfaces-configs
and restart ntp service in mgmt-vrf context.
management interface (eth0) can be configured using config interface eth0 ip add
command and removed using config interface eth0 ip remove command.
Requirement and design are covered in mgmt vrf design document.
Currently show command displays Linux command output, will update show
command display in next PR after concluding what would be the output for the
show commands. Added metric for default routes in dhcp and static, any changes
for metric will be addressed subsequently after discussing.

Signed-off-by: Harish Venkatraman <harish_venkatraman@dell.com>

**- What I did**
Added support for new feature management VRF using l3mdev.
**- How I did it**
Added commands to enable and disable management VRF using config commands and displays the management VRF information using show commands.
Added support to configure eth0 (management interface) ip address 

**- How to verify it**
Verified by running the below commands for configuring management VRF and delete management VRF
config vrf add mgmt
config vrf del mgmt
Use the below command to configure eth0 ip address.
config interface eth0 ip add ip/mask gatewayIP
Ex: config interface eth0 ip add 10.11.100.11/24 10.11.100.1
config interface eth0 ip remove ip/mask
config interface eth0 ip remove10.11.100.11/24
Verified NTP using below steps
- Added NTP_SERVERS in config_db.json
- Load config_db.json 
- restart ntp.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: build_debian.sh [hunk 1]
```diff
@@ -204,6 +204,7 @@ sudo LANG=C chroot $FILESYSTEM_ROOT apt-get -y install      \
 sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get -y install      \
     file                    \
     ifupdown2               \
+    ifmetric                \
```
Comment (lguohan) [inline build_debian.sh]: why do we need ifmetric?

File: files/image_config/interfaces/interfaces.j2 [hunk 2]
```diff
@@ -5,6 +5,11 @@
 # file: /etc/network/interfaces
 #
 {% endblock banner %}
+{% if (MGMT_VRF_CONFIG) and (MGMT_VRF_CONFIG['vrf_global']['mgmtVrfEnabled'] == "true") %}
+auto mgmt
+iface mgmt
+    vrf-table 1001
```
Comment (lguohan) [inline files/image_config/interfaces/interfaces.j2]: this is conflict with sonic-swss vrf assignment. Can you use 1000 instead?

https://github.com/Azure/sonic-swss/blob/master/cfgmgr/vrfmgr.cpp#L11

File: build_debian.sh (hunk 1 above)
Comment (vharish02) [reply build_debian.sh]: For setting the metric's for default and dhcp routes.

File: files/image_config/interfaces/interfaces.j2 (hunk 2 above)
Comment (vharish02) [reply files/image_config/interfaces/interfaces.j2]: Will use 1000

File: files/image_config/ntp/mgmtvrf_ntpq [hunk 3]
```diff
@@ -0,0 +1,5 @@
+ip link set dev lo master mgmt
```
Comment (lguohan) [inline files/image_config/ntp/mgmtvrf_ntpq]: what is the purpose for setting lo to master mgmt?

File: files/image_config/ntp/mgmtvrf_ntpq [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+sleep 1
```
Comment (lguohan) [inline files/image_config/ntp/mgmtvrf_ntpq]: why do we need this sleep?

File: files/image_config/ntp/mgmtvrf_ntpq (hunk 3 above)
Comment (kannankvs) [reply files/image_config/ntp/mgmtvrf_ntpq]: ntpq works only when loopback interface is part of mgmt vrf. Hence, loopback is temporarily added to mgmt vrf and moved back after executing the ntpq command.

File: files/image_config/ntp/mgmtvrf_ntpq (hunk 4 above)
Comment (kannankvs) [reply files/image_config/ntp/mgmtvrf_ntpq]: Immediately after attaching the lo to mgmt vrf,  when we call "ntpq", it fails few times (it succeeds few time). It is consistently successful only with this 1 second sleep.  If we can expect the user to execute the command couple of times to see successful result, this 1 second sleep is not required. If you want us to look into alternate solutions to solve this, let us know.

File: files/image_config/interfaces/interfaces.j2 [hunk 5]
```diff
@@ -26,25 +31,44 @@ auto eth0
 iface eth0 {{ 'inet' if prefix | ipv4 else 'inet6' }} static
     address {{ prefix | ip }}
     netmask {{ prefix | netmask if prefix | ipv4 else prefix | prefixlen }}
+{%     set vrf_table = 'default' %}
```
Comment (pavel-shirshov) [inline files/image_config/interfaces/interfaces.j2]: could it be part of else clause?

File: files/image_config/ntp/mgmtvrf_ntpq (hunk 3 above)
Comment (lguohan) [reply files/image_config/ntp/mgmtvrf_ntpq]: I found this is not intuitive. can you split the PR into two, one for management vrf only, without moving ntpd into management vrf and another one with ntpd move into the vrf. For the first PR, we still let ntp to run under default vrf, I think we are good and is able to approve.

File: files/image_config/interfaces/interfaces.j2 [hunk 6]
```diff
@@ -5,6 +5,11 @@
 # file: /etc/network/interfaces
 #
 {% endblock banner %}
+{% if (MGMT_VRF_CONFIG) and (MGMT_VRF_CONFIG['vrf_global']['mgmtVrfEnabled'] == "true") %}
+auto mgmt
+iface mgmt
+    vrf-table 4999
```
Comment (lguohan) [inline files/image_config/interfaces/interfaces.j2]: in the previous discussion, we ask this to be changed to 1000, why change this to 4999?

Comment (vharish02) [reply files/image_config/interfaces/interfaces.j2]: Guohan, while unit testing with either 1000 or any vrf table id less than 1000, we were getting the following error **"Error is : Nov 04 04:18:31 sonic networking[6426]: error: eth0: mgmt: vrf table id 999 out  of reserved range [1001,5000]**. Looking at the ifupdown2 code  the reserved range of vrf table id's are 1001 - 5000. So I changed it to 4999. Below is the snippet from the ifupdown2 code where they are initializing. 
    VRF_TABLE_START = 1001
    VRF_TABLE_END = 5000

**Comment by vharish02** [comment]
> suggest to split the pr into two, one for mvrf only, one for moving ntpd to mvrf.
will do

**Comment by vharish02** [comment]
> suggest to split the pr into two, one for mvrf only, one for moving ntpd to mvrf.

will do

File: files/image_config/interfaces/interfaces.j2 (hunk 6 above)
Comment (lguohan) [reply files/image_config/interfaces/interfaces.j2]: why not 5000 then?

**Comment by lguohan** [comment]
since ifupdown2 1.28 is installed, can you remove ifupdown2 from your PR?

File: files/image_config/interfaces/interfaces.j2 (hunk 6 above)
Comment (vharish02) [reply files/image_config/interfaces/interfaces.j2]: changed to 5000

**Comment by vharish02** [comment]
> since ifupdown2 1.28 is installed, can you remove ifupdown2 from your PR?

done

File: files/image_config/interfaces/interfaces.j2 [hunk 7]
```diff
@@ -5,6 +5,11 @@
 # file: /etc/network/interfaces
 #
 {% endblock banner %}
+{% if (MGMT_VRF_CONFIG) and (MGMT_VRF_CONFIG['vrf_global']['mgmtVrfEnabled'] == "true") %}
+auto mgmt
+iface mgmt
+    vrf-table 5000
```
Comment (nikos-github) [inline files/image_config/interfaces/interfaces.j2]: Isn't the table in the other PR 4999?

Comment (lguohan) [reply files/image_config/interfaces/interfaces.j2]: they need to modify the other pr accordingly.

**Comment by lguohan** [comment]
retest vsimage please

File: files/image_config/interfaces/interfaces.j2 (hunk 7 above)
Comment (vharish02) [reply files/image_config/interfaces/interfaces.j2]: > Isn't the table in the other PR 4999?

@nikos-github It's rebased now and shows 5000.

**Comment by batmancn** [comment]
I have few questions:
1. what is "l3mdev:mgmt" used for? Is it necessary?
2. what's metric used for?
3. where is "mgmt vrf design document" you mentioned?

**Pull request merged**